systemtest/apmservertest/filter.go

```diff
@@ -0,0 +1,116 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmservertest
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"encoding/json"
+	"io"
+
+	"go.elastic.co/apm/model"
+	"go.elastic.co/apm/transport"
+	"go.elastic.co/fastjson"
+)
+
+// TODO(axw) move EventMetadata and filteringTransport to go.elastic.co/apmtest,
+// generalising filteringTransport to work with arbitrary base transports. To do
+// that we would need to dynamically check for optional interfaces supported by
+// the base transport, and create passthrough methods.
+
+// EventMetadata holds event metadata.
+type EventMetadata struct {
+	System  *model.System
+	Process *model.Process
+	Service *model.Service
+	Labels  model.IfaceMap
+}
+
+// EventMetadata holds event metadata.
+type EventMetadataFilter interface {
```

Comment:
No strong feelings, but is a `Filter` supposed to manipulate data? The way I know filters is that they either let data go through or not but don't change the input data. What do you think about naming this something like `metadataComposer` or `metadataOperator` instead?

Reply:
I used this name to be in line with filter APIs available in a few of the agents, e.g. [Node.js](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#apm-add-filter), [Ruby](https://www.elastic.co/guide/en/apm/agent/ruby/current/api.html#api-agent-add-filter), [.NET](https://www.elastic.co/guide/en/apm/agent/dotnet/current/public-api.html#filter-api). I might otherwise have used "processor".

Reply:
understood; the code comment suggests to move parts of this to the go agent and make it reusable, therefore aligning the naming makes sense.